consul-core/src/main/java/com/smoketurner/dropwizard/consul/core/ConsulServiceListener.java

```diff
@@ -15,27 +15,64 @@
  */
 package com.smoketurner.dropwizard.consul.core;
 
+import com.orbitz.consul.ConsulException;
 import io.dropwizard.lifecycle.ServerLifecycleListener;
+import io.dropwizard.util.Duration;
 import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
 import org.eclipse.jetty.server.Server;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class ConsulServiceListener implements ServerLifecycleListener {
-
+  private static final Logger LOGGER = LoggerFactory.getLogger(ConsulServiceListener.class);
   private final ConsulAdvertiser advertiser;
+  private final Optional<Duration> retryInterval;
+  private final Optional<ScheduledExecutorService> scheduler;
 
   /**
    * Constructor
    *
    * @param advertiser Consul advertiser
+   * @param retryInterval When specified, will retry if service registration fails
    */
-  public ConsulServiceListener(final ConsulAdvertiser advertiser) {
+  public ConsulServiceListener(
+      final ConsulAdvertiser advertiser, @Nullable final Duration retryInterval) {
     this.advertiser = Objects.requireNonNull(advertiser);
+    this.retryInterval = Optional.ofNullable(retryInterval);
+    if (this.retryInterval.isPresent()) {
+      scheduler = Optional.of(Executors.newScheduledThreadPool(1));
+    } else {
+      scheduler = Optional.empty();
+    }
   }
 
   @Override
   public void serverStarted(final Server server) {
     final int applicationPort = getLocalPort(server);
     final int adminPort = getAdminPort(server);
-    advertiser.register(applicationPort, adminPort);
+    register(applicationPort, adminPort);
+  }
+
+  void register(int applicationPort, int adminPort) {
+    try {
+      advertiser.register(applicationPort, adminPort);
```

Comment:
@jplock , do you mean shutting down scheduler (if present) here?  when there is no exception meaning the service registration is completed successfully?

Reply:
no, I meant ensuring the scheduler is shutdown when shutting down the application

Reply:
Got it. Make sense. Will add it.